clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp

```diff
@@ -54,6 +55,7 @@ class MiscModule : public ClangTidyModule {
     CheckFactories.registerCheck<MisleadingIdentifierCheck>(
         "misc-misleading-identifier");
     CheckFactories.registerCheck<MisplacedConstCheck>("misc-misplaced-const");
+    CheckFactories.registerCheck<MustUseCheck>("misc-must-use");
```

Comment:
I do not like check name, it's to short, and doesn't say what is not used (member, var, global, function, class ?).
Second this should eb a bugprone check with preconfigured things like std::mutex, std::future, and probably few other.
Third i personally got check called bugprone-unused-local-non-trivial-variable, that is implemented like this:
```
void BugproneUnusedLocalNonTrivialVariableCheck::registerMatchers(MatchFinder *Finder) {
  Finder->addMatcher(varDecl(unless(isExpansionInSystemHeader()),
                             unless(isImplicit()),
                             unless(parmVarDecl()),
                             unless(decompositionDecl()),
                             unless(isExceptionVariable()),
                             hasLocalStorage(),
                             isDefinition(),
                             unless(hasType(namedDecl(matchesName(ExcludeTypeRegex)))),
                             unless(hasType(qualType(references(namedDecl(matchesName(ExcludeTypeRegex)))))),
                             decl().bind("var"),
                                 hasAncestor(functionDecl(unless(hasDescendant(declRefExpr(to(decl(equalsBoundNode("var"))))))))
                            ).bind("var"), this);
}

void BugproneUnusedLocalNonTrivialVariableCheck::check(const MatchFinder::MatchResult &Result) {
  const auto *MatchedDecl = Result.Nodes.getNodeAs<VarDecl>("var");

  auto type = MatchedDecl->getType();
  if (type.isTrivialType(*Result.Context) or type.isTriviallyCopyableType(*Result.Context)) return;

  diag(MatchedDecl->getLocation(), "unused local variable %0 of non trivial type %1, consider removing")
      << MatchedDecl << type;
}

It's not perfect, but what I'm thinking, is if this check and that shoudn't be a single check with inclusion/exclusion settings, so it could be configured to warn on all or just on some limited scope.
```

Reply:
Merging that check and mine STGM (as well as the naming of your check). Do you have source for your check you'd like to point to? Otherwise I'm happy to add in a lot of these.

Reply:
Is there a difference between isReferenced versus the following:

```
decl().bind("var"),         hasAncestor(functionDecl(unless(hasDescendant(declRefExpr(to(decl(equalsBoundNode("var"))))))))
```

Sorry still new to AST matchers

Reply:
I think isReferenced is newer, so stick to it. Fell free to merge those, check that i pointed out is old, so basicly the main thing that is missing in yuors is a validation of trivial type and support for inclusion/exclusion matchers.
and few tests, so just start with that. rename check and we go from this.

Reply:
SG added most of these checks, let me know if you think I should add any.